Use stateful dataloader to checkpoint data iteration order and token buffer

gokulavasan opened the pull request:
Summary: 

Use the stateful_dataloader from torchdata (https://github.com/pytorch/data/tree/main/torchdata/stateful_dataloader) for storing the token buffer and iteration data order. It requires a dependency on the nightly build of torchdata >= 20240426.

Also make sure the dataloader state has a different key per rank.

Test Plan:

Tested locally by first running 30 steps (checkpointing every 5 steps) and capturing all the loss values. Then deleting the last 3 checkpoints and then re-run the training and the loss values from step 16-30 match with what we had earlier in the first run. Note that this requires changes in the train.py to enable a deterministic run.

Reviewers: @tianyu-l 

Subscribers: @andrewkho 

Tasks:

Tags:

andrewkho commented on torchtitan/checkpoint.py:
```diff
@@ -20,6 +20,8 @@
     set_model_state_dict,
     set_optimizer_state_dict,
 )
+from torch.utils.data import DataLoader
```
Why do we need this too?

gokulavasan replied:
@andrewkho For typing. I can switch to StatefulDataLoader if we know that is going to be the one used in all scenarios

andrewkho commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -107,36 +108,71 @@ def __init__(
+
+        # Map-style dataset
+        # As skipping to the end throws an error, return an empty iterator
+        if skip_samples == len(self._data):
```
Does this happen at end-of-epoch? In this case should we just restart the iterator instead of returning empty iterator?

gokulavasan replied:
@andrewkho Yes, this is for end-of-epoch. I structured this method to take care of creating a skipped iterator. The caller takes care of restarting if self.infinite is set to true.

gokulavasan commented:
Looks like adding the index url (for torchdata) is causing other dependencies to not get installed. Will figure out how to fix this

fegin commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -107,36 +108,71 @@ def __init__(
+        self._all_tokens = state_dict["token_buffer"]
+
+    def state_dict(self):
+        return {"token_buffer": self._all_tokens, "sample_idx": self._sample_idx}
```
Is the `_all_tokens` a DTensor? If not, only rank0's dataloader data will be saved.

fegin replied:
Same as `_sample_idx`.

gokulavasan replied:
@fegin For now, the state_dict returned by the dataloader is keyed by the rankid. That is {0: state_dict}. Would that ensure the state of all the ranks is stored? Whether we want to use DTensor is something we are looking into (as resharding dataloader state is a challenge anyway so the benefit of storing it as DTensor is a bit limited)

fegin replied:
As long as the key is unique across ranks, using DTensor is not required.

tianyu-l commented on requirements.txt:
```diff
@@ -5,3 +5,5 @@ tensorboard
 sentencepiece
 tiktoken
 blobfile
+-i https://download.pytorch.org/whl/nightly/cpu
```
What is the earliest nightly that is good to test? Apr 25?

We depend on nightly pytorch in CI, so this can be removed when we land this PR.
https://github.com/pytorch/torchtitan/blob/main/.github/workflows/unit_test_4gpu.yaml#L36

gokulavasan replied:
@tianyu-l Need torchdata build from Apr 26 (20240426) at least

tianyu-l replied:
for HF datasets it seems from local test that we need `datasets >= 2.19.0` to use the `skip()` API

fegin replied:
confirm that I need to reinstall datasets to run with the PR

tianyu-l reviewed (commented):
First pass looks awesome already. Let me see if I can run some large scale experiments. What is the earliest nightly that's OK to run with this PR?

fegin commented:
@gokulavasan, @tianyu-l https://github.com/pytorch/pytorch/pull/125335 and https://github.com/pytorch/pytorch/pull/125334 should unblock this PR.

rlrs commented on torchtitan/checkpoint.py:
```diff
@@ -59,12 +61,30 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+            dist.get_rank() if (dist.is_available() and dist.is_initialized()) else 0
+        )
+
+    def state_dict(self) -> Dict[str, Any]:
+        if isinstance(self.dataloader, StatefulDataLoader):
+            return {self.rank_id: self.dataloader.state_dict()}
+        return {}
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        if isinstance(self.dataloader, StatefulDataLoader):
+            self.dataloader.load_state_dict(state_dict[self.rank_id])
```
This seems to break because `rank_id` gets cast to a string at some point. Should maybe cast it to a string manually in both functions?

rlrs commented:
I've been testing this out, and ran into an issue with resuming from a checkpoint. I suspect it's because of how `StatefulDataLoader` handles the state dict: https://github.com/pytorch/data/blob/11e16da61d7f5f587627c75e99ea664efef3e0f8/torchdata/stateful_dataloader/stateful_dataloader.py#L249

That is, a freshly initialized `StatefulDataLoader` does not have a state dict to load into? I'm not very familiar with how `DCP` works, so please correct me if it's wrong.

gokulavasan commented:
@rlrs Would it be possible to test it after my latest commit ([b9b045d](https://github.com/pytorch/torchtitan/pull/279/commits/b9b045d32933c2824ae6f667e944a51c3255a2d1)? I missed addin that part.

rlrs commented:
> @rlrs Would it be possible to test it after my latest commit ([b9b045d](https://github.com/pytorch/torchtitan/pull/279/commits/b9b045d32933c2824ae6f667e944a51c3255a2d1))? I missed adding that part.

I had already added that in my version. I can't get it to load the state_dict, unless I first call `iter(dataloader)` so that `dataloader._iterator is not None`. 

If I call `iter` before `DCP.load`, and then set `self._first_iter = True` in `HuggingFaceDataset.load_state_dict`, everything seems to work!

tianyu-l commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 torch >= 2.2.0.dev
```
I think we need to put the `torchdata` dependency here, maybe try
```suggestion
torch >= 2.2.0.dev
--find-links https://download.pytorch.org/whl/nightly/cpu/
torchdata >= 0.7.1.dev20240426+cpu
```
If works, we can remove the dependency in unit test workflow.

gokulavasan replied:
I attempted thus but it fails to find the nightly, instead installs the latest released version. We do plan to do a torchdata release soon (which will contain the latest StatefulDataLoader) and once that happens, we can remove the explicit pip installs in unit test workflows and main README.

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -99,32 +101,61 @@ def __init__(
 
         while True:
-            for sample in iter(self._data):
+            # Skip samples for the first iteration to resume from checkpoint
+            data_iter = (
+                self._get_skipped_data_iter(self._sample_idx)
+                if self._first_iter
+                else iter(self._data)
+            )
+            self._first_iter = False
```
Let's put all of them into the helper function to make it cleaner, e.g. with name `_get_data_iter` or `_get_maybe_skipped_data_iter`. Also this function doesn't need input, because `self._sample_idx` and `self._first_iter` are both class variables.

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -99,32 +101,61 @@ def __init__(
 
+    def _get_skipped_data_iter(self, skip_samples):
+        # As skipping to the end throws an error in case of map-style dataset, return an empty iterator
+        if skip_samples == len(self._data):
```
IIRC, `len(self._data)` is only available for map-style?

tianyu-l reviewed (commented):
Looks great! Had some minor comments.
For next steps, we can discuss if we should add a unit test to guard the correctness of checkpointable data loading, and the plan to migrate to DTensor-based checkpointing.

gokulavasan commented:
@tianyu-l Addressed PR comments (thank you!), added unit test, and made changes to the github workflows to allow running those unit tests. Let me know if the changes look okay. Regarding the move to DTensor, I think this requires analysis of what the benefits are (especially for storing unstructured state dict of dataloader).

If it is purely to reduce the replication of state across tensor/pipeline parallel groups, I think we can store the dataloader state just for the dp worker ranks (by using key as the dp_rank_id) and load it back instead of storing it for all global ranks. For now, with just the text tokens, this might not even be necessary as the state is not that big. Let me know how you would like to proceed.

gokulavasan commented:
@rlrs Thank you for your great analysis here (https://github.com/pytorch/torchtitan/pull/279#issuecomment-2104797493). 

Helped us narrow down the issue which basically boiled down to in-place loading of checkpoint of DCP. StatefulDataLoader doesn't currently return no state if dataloader iterator is not created while DCP expects the module it let it know what the keys the module is expecting. 

In order to get around this, I serialized the state of the dataloader and in this case there is only one key to load that is communicated by the DataLoaderWrapper to DCP - "<rank_id>".

tianyu-l commented on .github/workflows/unit_test_4gpu.yaml:
```diff
@@ -29,5 +29,7 @@ jobs:
 
         python -m pip install --force-reinstall --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121
+        python -m pip install --pre torchdata --index-url https://download.pytorch.org/whl/nightly/
+        pip install -e ".[dev]"
```
Do we need this?

gokulavasan replied:
Answered below though I am not sure what tests are run for the 4gpu case

tianyu-l replied:
ditto: Let's remove this `pip install -e ".[dev]"` if not necessary

tianyu-l commented on .github/workflows/unit_test_cpu.yaml:
```diff
@@ -23,4 +23,6 @@ jobs:
 
         pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121
+        pip install --pre torchdata --index-url https://download.pytorch.org/whl/nightly
+        pip install -e ".[dev]"
```
ditto

gokulavasan replied:
The pip install? I wasn't able to run the unit test without it, it says torchtitan package is not found

tianyu-l replied:
I just tried: it will work if you have `__init__.py` in the `test/datasets` subfolder. Alternatively you can just put the .py test file in `test/`

tianyu-l commented on torchtitan/checkpoint.py:
```diff
@@ -67,6 +70,33 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+                return
+
+            # Load state for the current rank
+            self.dataloader.load_state_dict(pickle.loads(state_dict[self.rank_id]))
```
wow is this made possible by recent change from StatefulDataloader? What nightly version can we use to run it?

gokulavasan replied:
Sorry what is made possible? This is just deserializing the byte string of dataloader state back to a state_dict and passing it in.

tianyu-l commented on test/datasets/test_dataset_checkpoint.py:
```diff
@@ -0,0 +1,56 @@
+from torchtitan.datasets.tokenizer import create_tokenizer
+
+
+class TestDatasetCheckpoint:
```
This test logic looks great! IIUC this is not in CI yet and needs to be run manually?

Also I think we need some sort of integration test as well -- basically save checkpoint at step T, and resume from checkpoint and assert we are loading normally at step T+1. In fact we are working on setting up examples to make it easier to add such a test. Let me check when can it land.

gokulavasan replied:
It is in the CI. In cpu unit test yaml file, I see pytest command that runs at the top level. It should pick up the test. When I ran it locally, here is what I see:

>test/datasets/test_dataset_checkpoint.py::TestDatasetCheckpoint::test_c4_resumption PASSED                                                                                                               [ 14%]
test/test_job_config.py::TestJobConfig::test_command_line_args PASSED                                                                                                                                    [ 28%]
test/test_job_config.py::TestJobConfig::test_job_config_file PASSED                                                                                                                                      [ 42%]
test/test_job_config.py::TestJobConfig::test_job_file_does_not_exist PASSED                                                                                                                              [ 57%]
test/test_job_config.py::TestJobConfig::test_empty_config_file PASSED                                                                                                                                    [ 71%]
test/test_job_config.py::TestJobConfig::test_job_config_file_cmd_overrides PASSED                                                                                                                        [ 85%]
test/test_job_config.py::TestJobConfig::test_print_help PASSED                                                                                                                                           [100%]

---------- coverage: platform linux, python 3.10.14-final-0 ----------
Coverage XML written to file coverage.xml

gokulavasan replied:
> Also I think we need some sort of integration test as well -- basically save checkpoint at step T, and resume from checkpoint and assert we are loading normally at step T+1. 

This would be nice but it might be outside the scope of this PR. The reason being that it this also requires setting up the run to be deterministic etc.

>  In fact we are working on setting up examples to make it easier to add such a test. Let me check when can it land.

Once this is available, we can add an integration test using that

tianyu-l reviewed (commented):
>  I think we can store the dataloader state just for the dp worker ranks (by using key as the dp_rank_id) and load it back instead of storing it for all global ranks.

This sounds quite good! @fegin would checkpointing behave in this expected way? I.e. if we use the same key for the same TP ranks, but different keys for different DP ranks, would it avoid saving extra copies and load correctly? If that's the case I agree we don't have to use DTensor for now.

gokulavasan commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -99,32 +101,63 @@ def __init__(
+        # Skip samples
+        if isinstance(self._data, IterableDataset):
+            it = iter(self._data)
+            # Naively iterate through the samples as skip may not be supported
```
Take a look at this notebook - https://colab.research.google.com/drive/15ZeB5yMBey00MJ9MOeK24R1xMn2Hp2AU#scrollTo=6weCenMx9ntp. Basically c4 dataset after split doesn't allow skip. This looks like a bug that we can post in HF github issues to find out more but for now naively iterating. I can maybe add a link to the github issue in the comment

tianyu-l replied:
As discussed offline, I can refactor it after this PR lands, because we'd like to use `skip()` for IterableDataset as well.

tianyu-l commented on .github/workflows/unit_test_4gpu.yaml:
```diff
@@ -31,5 +31,6 @@ jobs:
         pip config --user set global.progress_bar off
 
         python -m pip install --force-reinstall --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121
+        python -m pip install --pre torchdata --index-url https://download.pytorch.org/whl/nightly/
```
I think we need this in `.github/workflows/integration_test_periodic.yaml` as well.

Let's create an issue, tracking that we need to put torchdata in `requirements.txt` and `pyproject.toml` after the needed changes ship in an official release.

gokulavasan replied:
@tianyu-l Should pyproject.toml datasets dependency also enforce >= 2.19.0 version requirement?

tianyu-l replied:
ah yes I think so

gokulavasan replied:
Created https://github.com/pytorch/torchtitan/issues/351 to remove torchdata nightly pip install

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -99,32 +102,90 @@ def __init__(
+        return {"token_buffer": self._all_tokens, "sample_idx": self._sample_idx}
+
+
+class DpAwareDataLoader(StatefulDataLoader, Stateful):
```
Let's use `DPAwareDataloader`, because in the [PEP 8 guide](https://peps.python.org/pep-0008/)
> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

tianyu-l reviewed (approved):
Looks awesome! Thanks for the beautiful work!

Please address inline comments before merging.

lhoestq commented:
Hi ! I'm Quentin from HF :)
FYI we just added state_dict() and load_state_dict() in `datasets.IterableDataset`, which can resume iteration faster than just skipping samples !
